test/e2e/pkg/infra/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,44 @@
+package digitalocean
+
+import (
+	"context"
+	"fmt"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+	"github.com/tendermint/tendermint/test/e2e/pkg/infra"
+	e2essh "github.com/tendermint/tendermint/test/e2e/pkg/ssh"
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	sshPort     = 22
+	testappName = "testappd"
+)
+
+var _ infra.Provider = &Provider{}
+
+// Provider implements a docker-compose backed infrastructure provider.
+type Provider struct {
+	Testnet            *e2e.Testnet
+	InfrastructureData e2e.InfrastructureData
+	SSHConfig          *ssh.ClientConfig
+}
+
+// Noop currently. Setup is performed externally to the e2e test tool.
+func (p *Provider) Setup() error {
+	return nil
+}
+
+// Noop currently. Node creation is currently performed externally to the e2e test tool.
+func (p Provider) CreateNode(ctx context.Context, n *e2e.Node) error {
+	return nil
+}
+func (p Provider) StartTendermint(ctx context.Context, n *e2e.Node) error {
+	return e2essh.Exec(p.SSHConfig, fmt.Sprintf("%s:%d", n.IP, sshPort), fmt.Sprintf("systemctl start %s", testappName))
```

Comment:
How long does it roughly take to initiate the SSH connection here?

We can test out this SSH-based approach, but if the process manipulation operations take too long to execute they may interfere with the tests in unexpected ways.

If the SSH-based approach works too slowly we should consider adding some kind of proxy service to the nodes that can be accessed via HTTP and that has the appropriate privileges to restart the test app.

Reply:
The docker version of this uses port mapping with different ports for each process and all ports are exposed on localhost. I've added a new field `ExternalIP` that stores the reachable IP for the node and replaced the old `IP` field with `InternalIP` to specify the IP address internal to the network. It is difficult to harmonize the docker and DO version of this without running the process in the same network space as the docker-compose network space which would be a bit more complex change.